lib/TH/THGeneral.c

```diff
@@ -137,10 +159,17 @@ void* THRealloc(void *ptr, long size)
   if(size < 0)
     THError("$ Torch: invalid memory size -- maybe an overflow?");
 
-  ptr = realloc(ptr, size);
-  if(!ptr)
+  void *newptr = realloc(ptr, size);
+
+  if(!newptr && torchGCFunction) {
+    torchGCFunction(torchGCData);
+    newptr = realloc(ptr, size);
+  }
+
+  if(!newptr)
     THError("$ Torch: not enough memory: you tried to reallocate %dGB. Buy new RAM!", size/1073741824);
```

Comment:
Interesting. So what I'd like to understand is this. Is the following code correct?

    void * p = THAlloc(n);
    void * q = THRealloc(p, 2 * n);
    THFree(q);

Or should I have to write:

    void * p = THAlloc(n);
    void * q = THRealloc(p, 2 * n);
    THFree(q ? q : p);

Or is neither code correct because `THRealloc` never returns and I just can't free `p`?